qdax/core/emitters/cma_emitter.py

```diff
@@ -0,0 +1,374 @@
+from __future__ import annotations
+
+from functools import partial
+from typing import Optional, Tuple
+
+import jax
+import jax.numpy as jnp
+
+from qdax.core.cmaes import CMAES, CMAESState
+from qdax.core.containers.mapelites_repertoire import (
+    MapElitesRepertoire,
+    get_cells_indices,
+)
+from qdax.core.emitters.emitter import Emitter, EmitterState
+from qdax.types import Centroid, Descriptor, ExtraScores, Fitness, Genotype, RNGKey
+
+
+class CMAEmitterState(EmitterState):
+    """
+    Emitter state for the CMA-ME emitter.
+
+    Args:
+        random_key: a random key to handle stochastic operations. Used for
+            state update only, another key is used to emit. This might be
+            subject to refactoring discussions in the future.
+        cmaes_state: state of the underlying CMA-ES algorithm
+        previous_fitnesses: store last fitnesses of the repertoire. Used to
+            compute the improvment.
+        emit_count: count the number of emission events.
+    """
+
+    random_key: RNGKey
+    cmaes_state: CMAESState
+    previous_fitnesses: Fitness
+    emit_count: int
+
+
+class CMAEmitter(Emitter):
```

Comment:
If I understood correctly, this corresponds to the **improvement** Emitter. 

Thus I find it weird that: (1) this class is not called CMAImprovementEmitter, and (2) the other classes inherit from it.

Wouldn't it be better to have a base (abstract) class `CMAEmitter`, with `_ranking_criteria(...)` as abstract method, and that all the other CMA[...]Emitter inherit from it?

Reply:
Same remark for the `CMAEmitterState`

Reply:
I do agree for the Emitter!

For the EmitterState, the current design seems to make sense for me (as soon as the current CMAEmitter will have been renamed CMAImprovementEmitter).